gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java

```diff
@@ -54,6 +77,81 @@ public class GoogleCloudStorageClientImpl extends ForwardingGoogleCloudStorage {
             .setHttpRequestInitializer(httpRequestInitializer)
             .setDownscopedAccessTokenFn(downscopedAccessTokenFn)
             .build());
+    this.storageOptions = options;
+    this.storage =
+        clientLibraryStorage == null ? createStorage(credentials, options) : clientLibraryStorage;
+  }
+
+  @Override
+  public WritableByteChannel create(StorageResourceId resourceId, CreateObjectOptions options)
+      throws IOException {
+    logger.atFiner().log("create(%s)", resourceId);
+    checkArgument(
+        resourceId.isStorageObject(), "Expected full StorageObject id, got %s", resourceId);
+    // Update resourceId if generationId is missing
+    StorageResourceId resourceIdWithGeneration = resourceId;
+    if (!resourceId.hasGenerationId()) {
+      resourceIdWithGeneration =
+          new StorageResourceId(
+              resourceId.getBucketName(),
+              resourceId.getObjectName(),
+              getWriteGeneration(resourceId, options.isOverwriteExisting()));
+    }
+
+    GoogleCloudStorageClientWriteChannel channel =
+        new GoogleCloudStorageClientWriteChannel(
+            storage, storageOptions, resourceIdWithGeneration, options, backgroundTasksThreadPool);
+    channel.initialize();
+    return channel;
+  }
+
+  @Override
+  public void close() {
+    try {
+      try {
+        super.close();
+      } finally {
+        backgroundTasksThreadPool.shutdown();
+      }
+    } catch (Exception e) {
+      logger.atWarning().withCause(e).log("Exception while closing gcsClientImpl");
+    } finally {
+      backgroundTasksThreadPool = null;
+    }
+  }
+
+  /**
+   * Gets the object generation for a write operation
+   *
+   * <p>making getItemInfo call even if overwrite is disabled to fail fast in case file is existing.
+   *
+   * @param resourceId object for which generation info is requested
+   * @param overwrite whether existing object should be overwritten
+   * @return the generation of the object
+   * @throws IOException if the object already exists and cannot be overwritten
+   */
+  private long getWriteGeneration(StorageResourceId resourceId, boolean overwrite)
```

Comment:
Code duplication. Can we refactor and reuse?

Reply:
I don't see much scope for refactoring as extracting it to a separate helper class wouldn't add much value.